cluster-autoscaler/core/scale_down.go

```diff
@@ -103,6 +102,107 @@ func (n *NodeDeleteStatus) SetDeleteInProgress(status bool) {
 	n.deleteInProgress = status
 }
 
+type resourceName string
+type scaleDownResourcesLimits map[resourceName]int64
+type scaleDownResourcesDelta map[resourceName]int64
+
+// used as a value in scaleDownResourcesLimits if actual limit could not be obtained due to errors talking to cloud provider
+const limitUnknown = math.MinInt64
+
+func computeScaleDownResourcesLeftLimits(nodes []*apiv1.Node, resourceLimiter *cloudprovider.ResourceLimiter, timestamp time.Time) (scaleDownResourcesLimits, errors.AutoscalerError) {
+	totals := make(map[resourceName]int64)
+	totals[cloudprovider.ResourceNameCores], totals[cloudprovider.ResourceNameMemory] = calculateCoresAndMemoryTotal(nodes, timestamp)
+	resultScaleDownLimits := make(scaleDownResourcesLimits)
+
+	for _, resource := range resourceLimiter.GetResources() {
+		total := totals[resourceName(resource)]
+		min := resourceLimiter.GetMin(resource)
+
+		// we put only actual limits into final map. No entry means no limit.
+		if min > 0 {
+			if total > min {
+				resultScaleDownLimits[resourceName(resource)] = total - min
+			} else {
+				resultScaleDownLimits[resourceName(resource)] = 0
+			}
+		}
+	}
+	return resultScaleDownLimits, nil
+}
+
+func calculateCoresAndMemoryTotal(nodes []*apiv1.Node, timestamp time.Time) (int64, int64) {
+	var coresTotal, memoryTotal int64
+	for _, node := range nodes {
+		deleteTime, _ := deletetaint.GetToBeDeletedTime(node)
+		if deleteTime != nil && (timestamp.Sub(*deleteTime) < MaxCloudProviderNodeDeletionTime || timestamp.Sub(*deleteTime) < MaxKubernetesEmptyNodeDeletionTime) {
```

Comment:
side note: this is an ugly check that we probably need to do all over the place. Maybe we should add IsNodeBeingDeleted() function? Let's not do it in this PR though - it's large enough as is.

Reply:
+1